### PR TITLE
Patch/b1

### PR DESCRIPTION
**Pull request by anonymoussprocket**
## Breaking Changes
- `TezosNodeWriter.applyOperation` was renamed to `TezosNodeWriter.preapplyOperation`
- `TezosNodeWriter.testOperation` was renamed to `TezosNodeWriter.testContractInvocationOperation` and now estimates cost of contract invocation
- `TezosConseilClient.getBlock` now returns a single item, not an array
- `TezosConseilClient.getAccount` now returns a single item, not an array
- `TezosConseilClient.getOperationGroup` now returns a single item, not an array

## Fixes
- custom entry point hex encoding, thanks to @elkesrio

## Known issues
- Michelson lambda parameter encoding is not yet supported

## Future Breaking Changes
- `KeyStore` now contains `derivationPath`, in a subsequent release, functions that accept a `KeyStore` parameter and a `derivationPath` parameter will be modified to take only the former.

### CODE REVIEW

**Comment by vishakh** [comment]
Let us discuss changes to `KeyStore` before the next release. Any changes to the wallet file format should be very carefully reviewed.

File: src/chain/tezos/TezosMessageCodec.ts [hunk 1]
```diff
@@ -434,7 +434,7 @@ export namespace TezosMessageCodec {
                 hex += '04';
             } else {
                 hex += 'ff'
-                    + ('0' + (composite.entrypoint.length / 2).toString(16)).slice(-2)
+                    + ('0' + composite.entrypoint.length.toString(16)).slice(-2)
```
Comment (vishakh) [inline src/chain/tezos/TezosMessageCodec.ts]: Why did you have to make this change?

File: src/chain/tezos/TezosNodeWriter.ts [hunk 2]
```diff
@@ -226,8 +231,8 @@ export namespace TezosNodeWriter {
         const blockHead = await TezosNodeReader.getBlockHead(server);
         const forgedOperationGroup = forgeOperations(blockHead.hash, operations);
         const signedOpGroup = await signOperationGroup(forgedOperationGroup, keyStore, derivationPath);
-        const appliedOp = await applyOperation(server, blockHead.hash, blockHead.protocol, operations, signedOpGroup);
-        checkAppliedOperationResults(appliedOp);
+        const appliedOp = await preapplyOperation(server, blockHead.hash, blockHead.protocol, operations, signedOpGroup);
+
```
Comment (vishakh) [inline src/chain/tezos/TezosNodeWriter.ts]: Why are we removing this check?

File: src/identity/tezos/TezosLedgerWallet.ts [hunk 3]
```diff
@@ -39,15 +36,12 @@ export namespace TezosLedgerWallet {
     /**
      * Given a BIP44 derivation path for Tezos, get the Tezos Public Key
      * 
-     * @param derivationPath BIP44 Derivation Path
+     * @param derivationPath BIP32/44 derivation path
```
Comment (vishakh) [inline src/identity/tezos/TezosLedgerWallet.ts]: We should discourage BIP32 derivation paths.

File: src/reporting/tezos/TezosConseilClient.ts [hunk 4]
```diff
@@ -48,10 +48,11 @@ export namespace TezosConseilClient {
      * @param network Tezos network to query, mainnet, alphanet, etc.
      * @param hash Block hash to query for.
      */
-    export async function getBlock(serverInfo: ConseilServerInfo, network: string, hash: string): Promise<any[]> {
+    export async function getBlock(serverInfo: ConseilServerInfo, network: string, hash: string): Promise<any> {
         const query = ConseilQueryBuilder.setLimit(ConseilQueryBuilder.addPredicate(ConseilQueryBuilder.blankQuery(), 'hash', ConseilOperator.EQ, [hash], false), 1);
 
-        return getTezosEntityData(serverInfo, network, BLOCKS, query);
+        const r = await getTezosEntityData(serverInfo, network, BLOCKS, query);
+        return r[0];
```
Comment (vishakh) [inline src/reporting/tezos/TezosConseilClient.ts]: We should validate that at least one result was returned, else return an explicit error.

File: src/reporting/tezos/TezosConseilClient.ts [hunk 5]
```diff
@@ -77,7 +78,8 @@ export namespace TezosConseilClient {
     export async function getAccount(serverInfo: ConseilServerInfo, network: string, accountID: string): Promise<any[]> {
         const query = ConseilQueryBuilder.setLimit(ConseilQueryBuilder.addPredicate(ConseilQueryBuilder.blankQuery(), 'account_id', ConseilOperator.EQ, [accountID], false), 1);
 
-        return getTezosEntityData(serverInfo, network, ACCOUNTS, query);
+        const r = await getTezosEntityData(serverInfo, network, ACCOUNTS, query);
+        return r[0];
```
Comment (vishakh) [inline src/reporting/tezos/TezosConseilClient.ts]: We should validate that at least one result was returned, else return an explicit error.

File: src/reporting/tezos/TezosConseilClient.ts [hunk 6]
```diff
@@ -90,7 +92,8 @@ export namespace TezosConseilClient {
     export async function getOperationGroup(serverInfo: ConseilServerInfo, network: string, operationGroupID: string): Promise<any[]> {
         const query = ConseilQueryBuilder.setLimit(ConseilQueryBuilder.addPredicate(ConseilQueryBuilder.blankQuery(), 'hash', ConseilOperator.EQ, [operationGroupID], false), 1);
 
-        return getTezosEntityData(serverInfo, network, OPERATION_GROUPS, query);
+        const r = await getTezosEntityData(serverInfo, network, OPERATION_GROUPS, query);
+        return r[0];
```
Comment (vishakh) [inline src/reporting/tezos/TezosConseilClient.ts]: We should validate that at least one result was returned, else return an explicit error.

File: src/identity/tezos/TezosLedgerConnector.ts [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+export enum Instruction { // taken from https://github.com/obsidiansystems/ledger-app-tezos/blob/master/APDUs.md
+    INS_VERSION = 0x00,
+    INS_GET_PUBLIC_KEY = 0x02,
+    INS_PROMPT_PUBLIC_KEY = 0x03,
+    INS_SIGN = 0x04,
+    INS_SIGN_UNSAFE = 0x05
+}
+
+export default class TezosLedgerConnector {
```
Comment (vishakh) [inline src/identity/tezos/TezosLedgerConnector.ts]: Can we add a comment with the overall purpose of this class?

File: src/identity/tezos/TezosLedgerConnector.ts [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+    /**
+     * Get Tezos public key hash for a given BIP32/44 path and curve
+     * 
+     * @param {string} path BIP32/44 derivation path
+     * @param {boolean} prompt Prompt the user to provide the key on the hardware device, default true.
+     * @param {Curve} curve Curve to use for key generation, one of: ED25519 (default), SECP256K1, SECP256R1
+     */
+    async getAddress(path: string, prompt: boolean = true, curve: Curve = Curve.ED25519): Promise<string> {
+        if (!path.startsWith("44'/1729'")) {
```
Comment (vishakh) [inline src/identity/tezos/TezosLedgerConnector.ts]: Is there a reason for stopping people completely from using their own custom schemes?

File: src/identity/tezos/TezosLedgerConnector.ts [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+
+            const buffer = Buffer.alloc(part);
+            bytes.copy(buffer, 0, offset, offset + part);
+            message.push(buffer);
+        }
+
+        let response = await this.transport.send(0x80, instruction, 0x00, curve, message[0]);
+        for (let i = 1; i < message.length; i++) {
+            let code = (i === message.length - 1) ? 0x81 : 0x01;
+            response = await this.transport.send(0x80, instruction, code, curve, message[i]);
```
Comment (vishakh) [inline src/identity/tezos/TezosLedgerConnector.ts]: Is it possible to validate this response and throw an explicit error as needed?

File: src/identity/tezos/TezosLedgerConnector.ts [hunk 10]
```diff
@@ -0,0 +1,110 @@
+import Transport from "@ledgerhq/hw-transport";
```
Comment (vishakh) [inline src/identity/tezos/TezosLedgerConnector.ts]: How much of this file is being unit tested?

File: src/reporting/tezos/TezosConseilClient.ts (hunk 5 above)
Comment (anonymoussprocket) [reply src/reporting/tezos/TezosConseilClient.ts]: setLimit(1) will guarantee 0 or 1 results

File: src/identity/tezos/TezosLedgerConnector.ts (hunk 7 above)
Comment (anonymoussprocket) [reply src/identity/tezos/TezosLedgerConnector.ts]: this gets broken out into a module soon

File: src/identity/tezos/TezosLedgerConnector.ts (hunk 10 above)
Comment (anonymoussprocket) [reply src/identity/tezos/TezosLedgerConnector.ts]: i'm not immediately aware of a way to mock usb responses.

File: src/chain/tezos/TezosNodeWriter.ts (hunk 2 above)
Comment (anonymoussprocket) [reply src/chain/tezos/TezosNodeWriter.ts]: That's not a great error validation function, we should make a new one, that's already in progress

**Pull request merged**